src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,386 @@
+    super(storageLevel);
+
+    this.backoffConfig = backoffConfig;
+    this.project = project;
+    this.topic = topic;
+    this.subscription = subscription;
+    this.credentials = credentials;
+    this.autoAcknowledge = autoAcknowledge;
+  }
+
+  @Override
+  public void onStart() {
+    if (topic != null) {
+      createSubscription();
+    }
+    new Thread(this::receive).start();
+  }
+
+  @Override
+  public void onStop() {
+    //no-op
+  }
+
+  /**
+   * Create a new subscription (if needed) for the specified topic.
+   *
+   * @throws IllegalArgumentException when the specified Topic does not exists
+   * @throws RuntimeException         when the SubscriptionAdminClient cannot be created.
+   * @throws ApiException             when a non-retryable exception is thrown by the Pub/Sub client
+   */
+  protected void createSubscription() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+    int attempts = 5;
+
+    while (!isStopped() && attempts-- > 0) {
+
+      try (SubscriptionAdminClient subscriptionAdminClient = getSubscriptionAdminClient()) {
+
+        ProjectTopicName topicName = ProjectTopicName.of(project, topic);
+        ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(project, subscription);
+        int ackDeadline = 10; //10 seconds before resending the message.
+        subscriptionAdminClient.createSubscription(
+          subscriptionName, topicName, PushConfig.getDefaultInstance(), ackDeadline);
+        return;
+
+      } catch (ApiException ae) {
+        //If the subscription already exists, ignore the error.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.ALREADY_EXISTS)) {
+          return;
+        }
+
+        //This error is thrown is the Topic Name is not valid.
+        //Throw an Illegal Argument Exception so the pipeline fails.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.NOT_FOUND)) {
+          throw new IllegalArgumentException("Failed to create subscription. Topic Name is invalid.", ae);
+        }
+
+        //Retry if the exception is retriable.
+        if (ae.isRetryable()) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+          continue;
+        }
+
+        throw ae;
+      } catch (IOException ioe) {
+        throw new RuntimeException("Failed to create subscription.", ioe);
+      }
+
+    }
+
+    if (isStopped()) {
+      return;
+    }
+
+    throw new RuntimeException("Failed to create subscription after 5 attempts.");
+  }
+
+  /**
+   * Receive new messages and store based on the Storage Level settings.
+   *
+   * @throws ApiException     when the Pub/Sub API throws a non-retryable exception.
+   * @throws RuntimeException when the GrpcSubscriberStub cannot be created.
+   */
+  public void receive() {
+    SubscriberStubSettings subscriberStubSettings = getSubscriberStubSettings();
+
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Receiver Started execution");
```

Comment:
Just log it as debug without the if check. An `if` check is only necessary if the logging itself requires heavy computation (e.g. stringifying a complex object).